### ISSUE TITLE
Refactor monolithic training and evaluation loops for improved modularity

### ISSUE THREAD

**Issue by Jatewo** [SW10-Cryptanalysis/mamba#5]
The current codebase relies on large, monolithic functions that mix I/O logic, data processing, and training loops. We need to refactor these into a structured Object-Oriented (OO) framework. This will encapsulate state (like character offsets and model weights) and make the pipeline more extensible for different cipher types or model architectures.

## Architectural Goal
We will transition from procedural scripts to a class-based architecture where each component has a single responsibility.

### 1. `CipherTokenizer` (Core Logic)
* **Responsibilities:** Manage character-to-ID mappings and `char_offset`. 
* **Methods:** `encode(text)`, `decode(ids)`.
* **Benefit:** Centralizes the "source of truth" for the vocabulary, preventing index-out-of-range errors between training and inference.

### 2. `CipherDatasetManager` (Data & I/O)
* **Responsibilities:** Handle the abstraction of reading from nested ZIPs vs. JSON files on disk. 
* **Methods:** `get_dataloaders()`, `_scan_directory()`, `_load_file()`.
* **Benefit:** Simplifies `CipherDataset` by removing the complex I/O logic from the `__getitem__` method.

### 3. `MambaTrainer` (Training Engine)
* **Responsibilities:** Own the training state, including the model, optimizer, scheduler, and best validation loss.
* **Methods:** `train()`, `_run_epoch()`, `_save_checkpoint()`.
* **Benefit:** Removes global dependencies and allows for cleaner logging and early stopping implementation.

### 4. `CipherSolver` (Inference & Eval)
* **Responsibilities:** Encapsulate a trained model for "production" use.
* **Methods:** `load_from_checkpoint(path)`, `solve(ciphertext_ids)`, `evaluate_directory(path)`.
* **Benefit:** Allows us to use the model as a standalone tool without needing the training infrastructure.

---

## Proposed Class Structure

```python
class CipherTokenizer:
    """Handles alphabet mapping and token conversion."""
    def __init__(self, char_offset: int, plain_vocab_size: int): ...
    def encode(self, text: str) -> list[int]: ...
    def decode(self, ids: list[int]) -> str: ...

class CipherDatasetManager:
    """Discovers files and provides PyTorch DataLoaders."""
    def __init__(self, config: Config): ...
    def create_datasets(self) -> tuple[Dataset, Dataset]: ...
    def get_max_stats(self) -> tuple[int, int]: ...

class MambaTrainer:
    """Manages the optimization loop and model state."""
    def __init__(self, model: nn.Module, tokenizer: CipherTokenizer, config: Config): ...
    def fit(self, train_loader, val_loader): ...
    def _save_best_model(self, val_loss): ...

class CipherSolver:
    """High-level interface for inference and SER evaluation."""
    def __init__(self, checkpoint_path: str, tokenizer: CipherTokenizer): ...
    def decrypt(self, raw_cipher: list[int]) -> str: ...
    def run_benchmark(self, test_dir: Path): ...
